### PR TITLE
Transfer and claim automation now gets certificates in batches

### PR DESCRIPTION
**Pull request by tnickelsen**

### CODE REVIEW

File: domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs [hunk 1]
```diff
@@ -35,25 +38,38 @@ public async Task Run(CancellationToken stoppingToken)
...
-                        continue;
-                    }
+                        var response = await walletClient.GetGranularCertificates(subjectId, stoppingToken, limit: BatchSize, skip: numberOfFetchedCertificates);
 
-                    logger.LogInformation("Trying to claim {certificates} certificates for {subjectId}", response.Result.Count(), subjectId);
-                    var certificates = response.Result.OrderBy<GranularCertificate, int>(x => shuffle.Next()).ToList();
+                        if (response == null || !response.Result.Any())
+                        {
+                            logger.LogInformation("No certificates found for {subjectId}", subjectId);
+                            continue;
```
Comment (martinhenningjensen) [inline domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs:51]: Den her continue hører til while konstruktionen right? Giver det en uendelig løkke fordi hasMoreCertificates altid vil være sand?

Det kunne en test afdække :)

File: domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs [hunk 2]
```diff
@@ -35,25 +38,38 @@ public async Task Run(CancellationToken stoppingToken)
...
+                    var hasMoreCertificates = true;
+                    var numberOfFetchedCertificates = 0;
 
-                    if (response == null || !response.Result.Any())
+                    while (hasMoreCertificates)
                     {
-                        logger.LogInformation("No certificates found for {subjectId}", subjectId);
-                        continue;
-                    }
+                        var response = await walletClient.GetGranularCertificates(subjectId, stoppingToken, limit: BatchSize, skip: numberOfFetchedCertificates);
```
Comment (martinhenningjensen) [inline domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs:46]: Hvorfor skippe certificater, hvis sidste loop flyttede nogle certificater, så er de vel ikke i denne wallet mere?

File: domains/transfer/TransferAgreementAutomation/Worker/Service/ProjectOriginWalletService.cs [hunk 3]
```diff
@@ -28,32 +30,47 @@ ITransferAgreementAutomationMetrics metrics
...
+        var hasMoreCertificates = true;
+        var numberOfFetchedCertificates = 0;
 
-        logger.LogInformation("Found {certificatesCount} certificates to transfer for transfer agreement with id {id}", certificatesCount, transferAgreement.Id);
-
-        foreach (var certificate in certificates.Result)
+        while (hasMoreCertificates)
         {
-            if (!IsPeriodMatching(transferAgreement, certificate))
+            var certificates = await walletClient.GetGranularCertificates(transferAgreement.SenderId, new CancellationToken(), limit: BatchSize, skip: numberOfFetchedCertificates);
```
Comment (martinhenningjensen) [inline domains/transfer/TransferAgreementAutomation/Worker/Service/ProjectOriginWalletService.cs:40]: Same 'skip' comment as in claims worker

File: domains/transfer/TransferAgreementAutomation/Worker/Service/ProjectOriginWalletService.cs [hunk 4]
```diff
@@ -28,32 +30,47 @@ ITransferAgreementAutomationMetrics metrics
...
+
+            if (certificates == null)
             {
-                certificatesCount--;
+                logger.LogError("Something went wrong when getting certificates from the wallet for {senderId}. Response is null.", transferAgreement.SenderId);
```
Comment (martinhenningjensen) [inline domains/transfer/TransferAgreementAutomation/Worker/Service/ProjectOriginWalletService.cs:44]: Same 'continue' comment as in claims worker.

File: domains/transfer/Shared/ProjectOriginClients.Tests/ProjectOriginWalletClientTests.cs [hunk 5]
```diff
@@ -169,7 +169,21 @@ public async Task GetGranularCertificates()
         var walletClient = new ProjectOriginWalletClient(httpClient);
 
         //I cannot send any certificates to the wallet since I can't send to the registry first
-        var certsResponse = await walletClient.GetGranularCertificates(ownerSubject, new CancellationToken());
+        var certsResponse = await walletClient.GetGranularCertificates(ownerSubject, new CancellationToken(), limit: int.MaxValue, skip: 0);
```
Comment (martinhenningjensen) [inline domains/transfer/Shared/ProjectOriginClients.Tests/ProjectOriginWalletClientTests.cs:172]: Der kunne godt være en test med flere certificater end batch størrelsen... den håndtering er vist slet ikke testet. 

Hvad med edge cases hvor der er nøjagtigt batch-size certificater i wallet?

**Review by martinhenningjensen — CHANGES_REQUESTED** [review]

File: domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs (hunk 1 above)
Comment (tnickelsen) [reply domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs]: Har tilføjet unit tests

File: domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs (hunk 2 above)
Comment (tnickelsen) [reply domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs:45]: Fordi at flytte certifikater sker asynkront

File: domains/transfer/Shared/ProjectOriginClients.Tests/ProjectOriginWalletClientTests.cs (hunk 5 above)
Comment (tnickelsen) [reply domains/transfer/Shared/ProjectOriginClients.Tests/ProjectOriginWalletClientTests.cs:172]: Jeg har tilføjet tests. Lad mig vide, hvad du tænker

File: domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs [hunk 6]
```diff
@@ -35,25 +37,40 @@ public async Task Run(CancellationToken stoppingToken)
...
+                        }
 
-                    foreach (var cert in certificatesGrouped)
-                    {
-                        var productionCerts = cert.Where(x => x.CertificateType == CertificateType.Production).ToList();
-                        var consumptionCerts = cert.Where(x => x.CertificateType == CertificateType.Consumption).ToList();
-                        logger.LogInformation("Claiming {productionCerts} production certs and {consumptionCerts} consumption certs for {subjectId}", productionCerts.Count, consumptionCerts.Count, subjectId);
-                        await Claim(subjectId, consumptionCerts, productionCerts);
+                        numberOfFetchedCertificates += response.Result.Count();
+                        if (numberOfFetchedCertificates >= response.Metadata.Total)
```
Comment (martinhenningjensen) [inline domains/transfer/ClaimAutomation/Worker/Automation/ClaimService.cs:56]: Something is fishy here: I expect response.Metadata.Total to decrease with every call because certificates are claimed/moved. But numberOfFetchedCertificates is increased with each loop. When they cross each other, the loop excited, but more certificates might still exist.

The expression for more available certificates, should maybe be something along **if returned page size is less then returned total count**

**Review by martinhenningjensen — CHANGES_REQUESTED** [review]

File: domains/transfer/ClaimAutomation/Worker.UnitTests/Automation/ClaimServiceTests.cs [hunk 7]
```diff
@@ -13,18 +13,26 @@
 
 namespace Worker.UnitTests.Automation;
 
-public class ClaimServiceTests
+public class ClaimServiceTsts
```
[the remark this answers is not in the document]
Comment (tnickelsen) [reply domains/transfer/ClaimAutomation/Worker.UnitTests/Automation/ClaimServiceTests.cs]: Corrected

**Review by martinhenningjensen — APPROVED** [review]
